src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+        for (int i = 0; i < vocab_size; i++) {
+            token_scores[i] = TokenIdScore(i, logits[i]);
+        }
+
+        // top_k sampling
+        if (0 < parameters.top_k && parameters.top_k < (int)token_scores.size()) {
+            sampling_top_k(token_scores.data(),
+                           token_scores.data() + parameters.top_k,
+                           token_scores.data() + token_scores.size());
+            token_scores.resize(parameters.top_k);
+        }
+
+        // top_p sampling
+        if (0.f < parameters.top_p && parameters.top_p < 1.0f) {
+            auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+            token_scores.resize(pos - token_scores.data());
+        }
+
+        // sample next token
+        sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+        for (size_t i = 0; i < token_scores.size(); i++) {
+            logits[i] = token_scores[i].score;
+        }
+
+        thread_local std::random_device rd;
+        thread_local std::mt19937 gen(rd());
+
+        std::discrete_distribution<> dist(logits, logits + token_scores.size());
+        TokenIdScore out_token = token_scores[dist(gen)];
+
+        return out_token;
+    }
+};
+}  // namespace
+
+namespace ov {
+
+// todo: add batching
+ov::EncodedResults multinominal_decoding(ov::InferRequest& m_model_runner,
+                                         ov::Tensor input_ids,
+                                         ov::Tensor attention_mask,
+                                         ov::GenerationConfig generation_config,
+                                         std::shared_ptr<StreamerBase> streamer) {
+    ov::GenerationConfigHelper config_helper{generation_config};
+
+    ov::Shape prompts_shape = input_ids.get_shape();
+    size_t batch_size = prompts_shape[0];
+    size_t prompt_len = prompts_shape[1];
+
+    ov::EncodedResults results;
+    results.scores.resize(batch_size);
+    results.tokens.resize(batch_size);
+    std::fill(results.scores.begin(), results.scores.end(), 0);
+
+    // Initialize inputs
+    m_model_runner.set_tensor("input_ids", input_ids);
+    m_model_runner.set_tensor("attention_mask", attention_mask);
+
+    ov::Tensor position_ids = m_model_runner.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+
+    m_model_runner.get_tensor("beam_idx").set_shape({batch_size});
+    m_model_runner.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+
+    m_model_runner.infer();
+
+    auto logits_tensor = m_model_runner.get_tensor("logits");
+
+    int64_t sequence_offset = logits_tensor.get_shape().at(1) - 1;
+    size_t vocab_size = logits_tensor.get_shape().back();
+
+    float* logits = logits_tensor.data<float>() + (sequence_offset)*vocab_size;

```

Comment:
```suggestion
    float* logits = logits_tensor.data<float>() + sequence_offset * vocab_size;
```
This reads like a C style cast but you are just multipling to ints.

Reply:
applied